cmd/kubeadm/app/cmd/phases/certs.go

```diff
@@ -154,147 +86,138 @@ func getCertsSubCommands(defaultKubernetesVersion string) []*cobra.Command {
+	allCmd := &cobra.Command{
+		Use:     "all",
+		Short:   "Generates all PKI assets necessary to establish the control plane",
+		Long:    allCertsLongDesc,
+		Example: allCertsExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+			kubeadmutil.CheckErr(err)
+
+			err = certsphase.CreatePKIAssets(internalcfg)
+			kubeadmutil.CheckErr(err)
 		},
-		{
-			use:     "front-proxy-client",
-			short:   "Generates a front proxy CA client certificate and key for a Kubernetes cluster",
-			long:    frontProxyClientCertLongDesc,
-			cmdFunc: certsphase.CreateFrontProxyClientCertAndKeyFiles,
+	}
+	addFlags(allCmd, &cfgPath, cfg, true)
+
+	// SA creates the private/public key pair, which doesn't us x509 at all
+	saCmd := &cobra.Command{
+		Use:   "sa",
+		Short: "Generates a private key for signing service account tokens along with its public key",
+		Long:  saKeyLongDesc,
+		Run: func(cmd *cobra.Command, args []string) {
+			internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, cfg)
+			kubeadmutil.CheckErr(err)
+
+			err = certsphase.CreateServiceAccountKeyAndPublicKeyFiles(internalcfg)
+			kubeadmutil.CheckErr(err)
 		},
 	}
+	addFlags(saCmd, &cfgPath, cfg, false)
 
-	for _, properties := range subCmdProperties {
-		// Creates the UX Command
-		cmd := &cobra.Command{
-			Use:     properties.use,
-			Short:   properties.short,
-			Long:    properties.long,
-			Example: properties.examples,
-			Run:     runCmdFunc(properties.cmdFunc, &cfgPath, cfg, defaultKubernetesVersion),
-		}
+	subCmds := []*cobra.Command{allCmd, saCmd}
 
-		// Add flags to the command
-		cmd.Flags().StringVar(&cfgPath, "config", cfgPath, "Path to kubeadm config file. WARNING: Usage of a configuration file is experimental")
-		cmd.Flags().StringVar(&cfg.CertificatesDir, "cert-dir", cfg.CertificatesDir, "The path where to save the certificates")
-		if properties.use == "all" || properties.use == "apiserver" {
-			cmd.Flags().StringVar(&cfg.Networking.DNSDomain, "service-dns-domain", cfg.Networking.DNSDomain, "Alternative domain for services, to use for the API server serving cert")
-			cmd.Flags().StringVar(&cfg.Networking.ServiceSubnet, "service-cidr", cfg.Networking.ServiceSubnet, "Alternative range of IP address for service VIPs, from which derives the internal API server VIP that will be added to the API Server serving cert")
-			cmd.Flags().StringSliceVar(&cfg.APIServerCertSANs, "apiserver-cert-extra-sans", []string{}, "Optional extra altnames to use for the API server serving cert. Can be both IP addresses and DNS names")
-			cmd.Flags().StringVar(&cfg.API.AdvertiseAddress, "apiserver-advertise-address", cfg.API.AdvertiseAddress, "The IP address the API server is accessible on, to use for the API server serving cert")
-		}
+	certTree, err := certsphase.GetDefaultCertList().AsMap().CertTree()
+	kubeadmutil.CheckErr(err)
+
+	for ca, certList := range certTree {
```

Comment:
Please extract the anonymous functions into named functions. The globals that this is closing over need to be actual params to the function. The anonymous .Run functions are ok though.